### PR TITLE
Prepare v.0.3.0 release

### PR DESCRIPTION
**Pull request by ljvmiranda921**
## Tasklist (Deadline: August 10, 2018)
- [x] Run `black` and `flake8` for a clean format (drop)
- [x] Tests and minor changes for topologies (c/o @whzup )
- [x] Make a PyPI friendly README by following this [link](https://packaging.python.org/guides/making-a-pypi-friendly-readme/)
- [x] Boundary conditions #150 (c/o @jolayfield) *Moved to v.0.4.0 or on dev branch*
- [x] Resolve issue with PyYaml, you can find a related issue in yaml/pyyaml#207
- [x] Update README (update authors and maintainers)
- [x] Merge all dependency updates in #162, #166, #173, and #178 
- [x] Resolve all merge conflicts
- [x] Update HISTORY.rst 
- [x] Write Release Notes
- [x] Bump version and update setup.py
- [x] **Release:** merge to `master`, make a Release, and upload to PyPI

*Please write the changelog below for our release notes. It would be nice to follow [this](https://github.com/ljvmiranda921/pyswarms/releases/tag/v.0.2.0) link.*

---
# Release v.0.3.0

We're proud to present the release of PySwarms version 0.3.0! Coinciding with this, we would like to welcome Aaron Moser (@whzup) as one of the project's maintainers! v.0.3.0 includes new topologies, a `static` option to configure a particle's neighbor/s, and a revamped `plotters` module. We would like to thank our contributors for helping us with this release.

## Release notes

- **NEW:** More basic particle topologies in the `pyswarms.backend` module - #142, #151, #155, #177 
- **NEW:** Ability to make topologies static or dynamic - #164 
- **NEW:** A `GeneralOptimizerPSO` class. The `GeneralOptimizerPSO` class has an additional attribute for the topology used in the optimization - #151
- **NEW:** A `plotters` module for swarm visualization. The `environments` module is now deprecated - #135, #172 
- **FIX:** Bugfix for optimizations not returning the best cost - #176 
- **FIX:** Bugfix for `setup.py` not running on Windows - #175
- **IMPROVED:** Objective functions can now be parametrized. Helpful for your custom-objective functions - #144. Thanks, @bradahoward!
- **IMPROVED:** New single-objective functions - #168. Awesome work, @jayspeidell! 

## New Topologies and the `GeneralOptimizerPSO` Class

New topologies were added to improve the ability to customize how a swarm behaves during optimization. In addition, a `GeneralOptimizerPSO` class was added to enable switching-out various topologies. Check out the description below!

### New Topology classes and the `static attribute`

The newly added topologies expand on the existing ones (Star and Ring topology) and increase the built-in variety of possibilities for users that want to build their custom swarm implementation from the `pyswarms.backend` module. The new topologies include:
    - `Pyramid` topology: Computes the neighbours using a Delaunay triangulation of the particles.
    - `Random` topology: Computes the neighbours randomly, but systematically.
    - `VonNeumann` topology: Computes the neighbours using a Von Neumann topology (inherited from the Ring topology)
With these new topologies, the ability to change the behaviour of the topologies was added in form of a `static` argument that is passed when initializing a `Topology` class. The `static` parameter is a boolean that decides whether the neighbours in the topologies are computed every iteration (`static=False`) or only in the first one (`static=True`). It is passed as a parameter at the initialization of the topology and is `False` by default. Additionally, the `LocalBestPSO` now also takes a `static` parameter to pass this information to its Ring topology. For an example see below.

### The `GeneralOptimizerPSO` class

The new topologies can also be easily used in the new `GeneralOptimizerPSO` class which extends the collection of optimizers. In addition to the parameters used in the `GlobalBestPSO` and `LocalBestPSO` classes, the `GeneralOptimizerPSO` uses a `topology` argument. This argument passes a `Topology` class to the `GeneralOptimizerPSO`.

```python
from pyswarms.single import GeneralOptimizer
from pyswarms.backend.topology import Random

options = {"w": 1, "c1": 0.4, "c2": 0.5, "k": 3}
topology = Random(static=True)
optimizer = GeneralOptimizerPSO(n_particles=20, dimensions=4, options=options, bounds=bounds, topology=topology)
```
## The `plotters` module

The environments module is now deprecated. Instead, we have a plotters module that takes a property of the optimizer and plots it with minimal effort. The whole module is built on top of `matplotlib`.

```python
 import pyswarms as ps
 from pyswarms.utils.functions import single_obj as fx
 from pyswarms.utils.plotters import plot_cost_history

 # Set-up optimizer
 options = {'c1':0.5, 'c2':0.3, 'w':0.9}
 optimizer = ps.single.GlobalBestPSO(n_particles=50, dimensions=2, options=options)
 optimizer.optimize(fx.sphere_func, iters=100)

 # Plot the cost
 plot_cost_history(optimizer.cost_history)
 plt.show()
```
Imgur

We can also plot the animation...
 
 ```python
 from pyswarms.utils.plotters.formatters import Mesher
 from pyswarms.utils.plotters.formatters import Designer
 from pyswarms.utils.plotters import plot_contour, plot_surface

 # Plot the sphere function's mesh for better plots
 m = Mesher(func=fx.sphere_func)

 # Adjust figure limits
 d = Designer(limits=[(-1,1), (-1,1), (-0.1,1)],
              label=['x-axis', 'y-axis', 'z-axis'])
 ```
 
 In 2D,
 
 ```python
 plot_contour(pos_history=optimizer.pos_history, mesher=m, mark=(0,0))
 ```
 
 Contour
 
 Or in 3D!
 
 ```python
 pos_history_3d = m.compute_history_3d(optimizer.pos_history) # preprocessing
 animation3d = plot_surface(pos_history=pos_history_3d,
                            mesher=m, designer=d,
                            mark=(0,0,0))    
 ```
 
Surface

### CODE REVIEW

**Comment by whzup** [comment]
I can't push changes into this PR. So you have to do the `black` and `flake8` runs 😄

File: tests/optimizers/conftest.py [hunk 1]
```diff
@@ -8,63 +8,109 @@
...
+                Random(static=False), Random(static=True)
+                ])
+def topology(request):
+    """Parametrized topology parameter"""
+    topology_ = request.param
+    return topology_
+
+
+@pytest.fixture(scope="module", params=[True, False])
+def static(request):
```
Comment (whzup) [inline tests/optimizers/conftest.py]: To be consistent we should replace this fixture with parametrization marks.

File: tests/optimizers/test_general_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        {"c2": 0.7, "w": 0.5, "k": 2},
+        {"c1": 0.5, "w": 0.5, "k": 2},
+        {"c1": 0.5, "c2": 0.7, "k": 2},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5},
+    ],
+)
+def test_keyword_exception_random(options):
+    """Tests if exceptions are thrown when keywords are missing and a Random topology is chosen"""
+    with pytest.raises(KeyError):
+        GeneralOptimizerPSO(5, 2, options, Random())
```
Comment (whzup) [inline tests/optimizers/test_general_optimizer.py]: Add a `static` parameter.

File: tests/optimizers/test_general_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": -1},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 6},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 0.5}
+    ],
+)
+def test_invalid_k_value(options):
+    """Tests if exception is thrown when passing
+    an invalid value for k when using a Random topology"""
+    with pytest.raises(ValueError):
+        GeneralOptimizerPSO(5, 2, options, Random())
```
Comment (whzup) [inline tests/optimizers/test_general_optimizer.py]: `static` parameter

File: tests/optimizers/test_general_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": -1, "p": 2},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 6, "p": 2},
+        {"c1": 0.5, "c2": 0.7, "w": 0.5, "k": 2, "p": 5},
+    ],
+)
+def test_invalid_k_or_p_values(options, static):
+    """Tests if exception is thrown when passing
+    an invalid value for k or p when using a Ring topology"""
+    with pytest.raises(ValueError):
+        GeneralOptimizerPSO(5, 2, options, Ring(static=static))
```
Comment (whzup) [inline tests/optimizers/test_general_optimizer.py]: `VonNeumann`

File: tests/optimizers/test_general_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+)
+def test_bounds_size_exception(bounds, options, topology):
+    """Tests if exceptions are raised when bound sizes are wrong"""
+    with pytest.raises(IndexError):
+        GeneralOptimizerPSO(5, 2, options=options, topology=topology, bounds=bounds)
+
+
+@pytest.mark.parametrize(
+    "topology",
+    [Star(), Ring(), Pyramid(), Random()]
```
Comment (whzup) [inline tests/optimizers/test_general_optimizer.py]: Maybe we should use the `static` versions in this and the following tests as well.
It would also work if we create a variable:
```python
topologies = [Star(), Ring(), Ring(static=True), etc.]
```

And then use it in the decorator like so:
```python
@pytest.mark.parametrize("topology", topologies)
```

It would spare a lot of copy-pasting.

**Comment by ljvmiranda921** [comment]
Yesss, if you want to make changes, just do another PR again on this branch. Once merged, the diffs will reflect in this PR.

File: tests/optimizers/test_general_optimizer.py (hunk 5 above)
Comment (ljvmiranda921) [reply tests/optimizers/test_general_optimizer.py]: Yes that’s correct

File: tests/backend/conftest.py [hunk 6]
```diff
@@ -10,17 +10,18 @@
 # Import from package
 from pyswarms.backend.swarms import Swarm
 
+
 @pytest.fixture
 def swarm():
```
[the remark this answers is not in the document]
Comment (ljvmiranda921) [reply tests/backend/conftest.py]: You can make another one if it does not fit your use-case

File: pyswarms/single/general_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        # Initialize the resettable attributes
+        self.reset()
+        # Initialize the topology and check for type
+        if not isinstance(topology, Topology):
+            raise TypeError("Parameter `topology` must be a Topology object")
+        else:
+            self.top = topology
+
+        # Case for the Ring topology
+        if isinstance(topology, Ring):
```
[the remark this answers is not in the document]
Comment (ljvmiranda921) [reply pyswarms/single/general_optimizer.py]: Hmmm, in v.0.4.0, let’s make an errors module so that our assert statements and checks are abstracted. 

Better make VonNeumann explicit here just to be sure

**Comment by ljvmiranda921** [comment]
Hi @whzup , you can make changes on this branch by submitting a PR (just the same thing you've been doing before). Once merged, the changes will reflect here. I suggest that you take over the changes especially for the features you have introduced in this version (`VonNeumann`, `static` parameter, tests, etc.). :+1:  Make sure to cover everything in your PR, so that we won't make multiple commits on the same diffs.

File: tests/optimizers/conftest.py (hunk 1 above)
Comment (ljvmiranda921) [reply tests/optimizers/conftest.py]: I wonder why this showed up again, didn't I comment on this before? Or is it in a different `conftest`?

Comment (whzup) [reply tests/optimizers/conftest.py]: It's in a different `conftest` 😄

File: tests/optimizers/test_general_optimizer.py (hunk 5 above)
Comment (whzup) [reply tests/optimizers/test_general_optimizer.py]: I noticed that we already have a parametrized fixture for all the topologies so no need for the variable 👍

**Comment by ljvmiranda921** [comment]
If in two weeks the pyyaml bug hasn't been fixed, then let's just ignore this.

**Comment by ljvmiranda921** [comment]
## Note on Requirements

Reference: #162, #166, #173, #178
Let's merge all features first then work on updating the requirements file. So the merge procedure will look like this:
- Merge all features from development branch.
- Close and delete pyup branches (pyup will automatically create new PRs without conflicts)
- Create a new branch, `update-requirements` (even with the new `master` with v.0.3.0 features)
- Merge all new PRs to `update-requirements`. Update `setup.py` if possible. 
- Merge `update-requirements` to `master`.
- Upload to PyPI and create a v.0.3.0 release.

**Comment by whzup** [comment]
I had `black`  on 88 lines. It still looked ugly 😕 Let's leave that one out.

**Comment by ljvmiranda921** [comment]
Yes, I think you're fine and it's my turn to finish all other admin tasks. Let me do them next week, I'll now move between cities this weekend! :+1: 

For the Changelog, you can just copy paste them below the TaskList above. :+1:

**Comment by ljvmiranda921** [comment]
Awesome work, can you also link the corresponding PR/s for each feature? Thank you so much and sorry for the hassle.

**Comment by ljvmiranda921** [comment]
## Release Status

Everything is fine on my end. We are ready to release on August 10, 2018.
If there are any bugs before that time, hopefully we can resolve them. If not,
let's just make patch releases (e.g. 0.3.1, 0.3.2, etc.) in the future.

For boundary conditions, let's just put that on v.0.4.0. It coincides with other features
we are actually planning so it might be a better fit.

I think the Release Notes are fine. We can even edit that after releasing so there's no pressure
to make it perfect from the get-go.

**Comment by whzup** [comment]
How did you make the fancy pictures in the release notes for 0.2.0? Maybe we could use something similar for this one 😋 BTW, have you updated [this part of the documentation](https://pyswarms.readthedocs.io/en/master/dev.api.html) with the new features?

**Comment by whzup** [comment]
I added this
```python
from pyswarms.utils.plotters import plot_contour, plot_surface
```
to the changelog for completeness 👍

**Comment by whzup** [comment]
These parts of the documentation are outdated:
- https://pyswarms.readthedocs.io/en/master/dev.optimizer.html#writing-unit-tests We don't use the `unittest` module anymore
- https://pyswarms.readthedocs.io/en/master/api/pyswarms.base.html#module-pyswarms.base.base_single We should include the `GeneralOptimizerPSO` class in the "See also"

**Comment by ljvmiranda921** [comment]
Hi @whzup ,

> How did you make the fancy pictures in the release notes for 0.2.0?

LaTeX with TikZ :+1: 

>  Maybe we could use something similar for this one

What kind of diagram do you propose?

> BTW, have you updated [this part of the documentation](https://pyswarms.readthedocs.io/en/master/dev.api.html) with the new features?

I think there's no need. This part only describes the general API. And our new version doesn't really break that.

**Comment by ljvmiranda921** [comment]
Hi @whzup ,

> These parts of the documentation are outdated:

Would you like to make a PR for this? I can attend to this next week though :confused:

**Comment by whzup** [comment]
Hey @ljvmiranda921, I was a bit inactive the last few days due to some issues with my computer and time management 😄 I started working on the deprecated parts of the docs and I thought while on it, I could integrate the roadmap? We could put it in the General section, what do you think?

**Comment by ljvmiranda921** [comment]
> I started working on the deprecated parts of the docs and I thought while on it, I could integrate the roadmap? We could put it in the General section, what do you think?

Oh, you mean put the roadmap in the documentation? Sure I think that's a nice idea but our [roadmap](https://github.com/ljvmiranda921/pyswarms/wiki/Development-Roadmap) seems to be a "live" document that we'll keep on updating. Normally I'll start with something general and then we make it more specific etc. 

But yeah sure, we can put a "Roadmap" section under general (last part) and then we update them after every release!

**Comment by whzup** [comment]
Hi @ljvmiranda921, I just saw the [release notes for GitHub Desktop](https://desktop.github.com/release-notes/) and I really like their formatting. We could use some of their elements, for example:

- **NEW**: Basic particle topologies in the pyswarms.backend module - #142, #155, #176, #177
- **NEW**: Ability to make the topologies static or dynamic - #164

I guess the labels (new, fixed, etc.) are made with images.

**Comment by ljvmiranda921** [comment]
This looks interesting. Wanna do it for v.0.3.0?  I think we can get by with bold labels. :+1:

**Comment by whzup** [comment]
I changed the release notes above. They look pretty slick now, don't they?😋

**Comment by whzup** [comment]
I do have access! Shall I do this today? I'd have some spare time 😄 For the History we could use coloured labels like the GitHub Desktop example, what do you think? If so, we should style them differently though 😋

**Comment by ljvmiranda921** [comment]
> Shall I do this today? I'd have some spare time smile

Sure! If you want, then much better!

> For the History we could use coloured labels like the GitHub Desktop example, what do you think? If so, we should style them differently though

How can we achieve this? Small image labels or just colored text? If the icons look ugly then using bold letters may suffice. :+1:

**Comment by whzup** [comment]
@ljvmiranda921, I reformatted the [release notes on GitHub](https://github.com/ljvmiranda921/pyswarms/releases). In my opinion, they look pretty slick now 😄 I unified the formatting and corrected some spelling and grammar errors.

**Comment by ljvmiranda921** [comment]
Understood. Yes, they look much cleaner now!

**Comment by ljvmiranda921** [comment]
Hi @whzup , this looks good already. I will merge to `master` today so that tomorrow I can just upload the tarballs and wheel files and then we'll be all set!
